### PR TITLE
03621 pause/unpause token precompile

### LINKED ISSUE #3621: HIP 514: Add token management functions for pause functions
### Problem

The hts precompile flow is missing support for the pause flow

### Solution

Add support for the pause flow

- `pauseToken(address)`
- `unPauseToken(address)`

### Alternatives

_No response_

### PR DESCRIPTION
**Pull request by ivanparnarev**
Signed-off-by: ivanparnarev <ivan.parnarev@limechain.tech>

**Description**:

**Related issue(s)**:

Fixes #3621 

**Notes for reviewer**:

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [ ] Tested (unit, integration, etc.)

### CODE REVIEW

File: hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/AbiConstants.java [hunk 1]
```diff
@@ -1,119 +1,120 @@
...
+    // getApproved(uint256 tokenId)
+    public static final int ABI_ID_ERC_GET_APPROVED = 0x081812fc;
+    // isApprovedForAll(address owner, address operator)
+    public static final int ABI_ID_ERC_IS_APPROVED_FOR_ALL = 0xe985e9c5;
+    // ownerOf(uint256 tokenId)
+    public static final int ABI_ID_ERC_OWNER_OF_NFT = 0x6352211e;
+    // tokenURI(uint256 tokenId)
+    public static final int ABI_ID_ERC_TOKEN_URI_NFT = 0xc87b56dd;
+
+    // pauseToken(address)
```
Comment (IvanKavaldzhiev) [inline hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/AbiConstants.java:85]: To be consistent with the other comments you can add argument name : `pauseToken(address token)`

File: hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/codec/DecodingFacade.java [hunk 2]
```diff
@@ -38,765 +35,977 @@
...
+    private static final Function ERC_TRANSFER_FROM_FUNCTION =
+            new Function("transferFrom(address,address,uint256)");
+    private static final Bytes ERC_TRANSFER_FROM_SELECTOR =
+            Bytes.wrap(ERC_TRANSFER_FROM_FUNCTION.selector());
+    private static final ABIType<Tuple> ERC_TRANSFER_FROM_DECODER =
+            TypeFactory.create("(bytes32,bytes32,uint256)");
+
+    private static final Function PAUSE_TOKEN_FUNCTION =
+            new Function("pauseToken(address)", INT_OUTPUT);
+
```
Comment (IvanKavaldzhiev) [inline hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/codec/DecodingFacade.java:175]: Remove white line so that `PAUSE_TOKEN` variables are structured together

File: hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/codec/DecodingFacade.java [hunk 3]
```diff
@@ -38,765 +35,977 @@
...
+    private static final Bytes PAUSE_TOKEN_SELECTOR = Bytes.wrap(PAUSE_TOKEN_FUNCTION.selector());
+
```
Comment (IvanKavaldzhiev) [inline hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/codec/DecodingFacade.java:177]: Remove white line so that `PAUSE_TOKEN` variables are structured together

File: hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/codec/DecodingFacade.java [hunk 4]
```diff
@@ -38,765 +35,977 @@
...
+    private static final ABIType<Tuple> PAUSE_TOKEN_DECODER = TypeFactory.create("(bytes32)");
+
+    private static final Function UNPAUSE_TOKEN_FUNCTION =
+            new Function("unpauseToken(address)", INT_OUTPUT);
+
```
Comment (IvanKavaldzhiev) [inline hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/codec/DecodingFacade.java:182]: Remove white line so that `UNPAUSE_TOKEN` variables are structured together

File: hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/codec/PauseWrapper.java [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+ * limitations under the License.
+ * ‍
+ *
+ */
+
+import com.hederahashgraph.api.proto.java.TokenID;
+
+public record PauseWrapper(TokenID token) {
+
+    public static PauseWrapper pause(final TokenID token) {
```
Comment (IvanKavaldzhiev) [inline hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/codec/PauseWrapper.java:44]: Remove this method since it's confusing. We don't want to actually `pause` anything but assign `TokenID` value. You can use directly the constructor generated from the `record` type.

File: hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/codec/UnpauseWrapper.java [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+ * limitations under the License.
+ * ‍
+ *
+ */
+
+import com.hederahashgraph.api.proto.java.TokenID;
+
+public record UnpauseWrapper(TokenID token) {
+
+    public static UnpauseWrapper unpause(final TokenID token) {
```
Comment (IvanKavaldzhiev) [inline hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/codec/UnpauseWrapper.java:44]: Remove this method since it's confusing. We don't want to actually `unpause` anything but assign `TokenID` value. You can use directly the constructor generated from the record type.

File: test-clients/src/main/java/com/hedera/services/bdd/suites/contract/precompile/PauseUnpauseTokenAccountPrecompileSuite.java [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+                                .exposingCreatedIdTo(id -> tokenID.set(asToken(id))),
+                        uploadInitCode(PAUSE_UNPAUSE_CONTRACT),
+                        contractCreate(PAUSE_UNPAUSE_CONTRACT))
+                .when(
+                        withOpContext(
+                                (spec, opLog) ->
+                                        allRunFor(
+                                                spec,
+                                                contractCreate(PAUSE_UNPAUSE_CONTRACT)
+                                                        .bytecode(PAUSE_UNPAUSE_CONTRACT),
```
Comment (IvanKavaldzhiev) [inline test-clients/src/main/java/com/hedera/services/bdd/suites/contract/precompile/PauseUnpauseTokenAccountPrecompileSuite.java:116]: It seems you have created `PAUSE_UNPAUSE_CONTRACT` twice.

File: test-clients/src/main/java/com/hedera/services/bdd/suites/contract/precompile/PauseUnpauseTokenAccountPrecompileSuite.java [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
```
Comment (IvanKavaldzhiev) [inline test-clients/src/main/java/com/hedera/services/bdd/suites/contract/precompile/PauseUnpauseTokenAccountPrecompileSuite.java:148]: It seems you have created `PAUSE_UNPAUSE_CONTRACT` twice.

File: test-clients/src/main/java/com/hedera/services/bdd/suites/contract/precompile/PauseUnpauseTokenAccountPrecompileSuite.java [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
```
Comment (IvanKavaldzhiev) [inline test-clients/src/main/java/com/hedera/services/bdd/suites/contract/precompile/PauseUnpauseTokenAccountPrecompileSuite.java:180]: It seems you have created `PAUSE_UNPAUSE_CONTRACT` twice.

File: test-clients/src/main/java/com/hedera/services/bdd/suites/contract/precompile/PauseUnpauseTokenAccountPrecompileSuite.java [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+                                .initialSupply(0)
+                                .exposingCreatedIdTo(id -> tokenID.set(asToken(id))),
+                        uploadInitCode(PAUSE_UNPAUSE_CONTRACT),
+                        contractCreate(PAUSE_UNPAUSE_CONTRACT))
+                .when(
+                        withOpContext(
+                                (spec, opLog) ->
+                                        allRunFor(
+                                                spec,
+                                                contractCreate(PAUSE_UNPAUSE_CONTRACT)
```
Comment (IvanKavaldzhiev) [inline test-clients/src/main/java/com/hedera/services/bdd/suites/contract/precompile/PauseUnpauseTokenAccountPrecompileSuite.java:211]: It seems you have created `PAUSE_UNPAUSE_CONTRACT` twice.

File: hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/SyntheticTxnFactory.java [hunk 11]
```diff
@@ -420,6 +424,22 @@ public TransactionBody.Builder nodeStakeUpdate(
...
+
+        return TransactionBody.newBuilder().setTokenPause(builder);
+    }
+
+    public TransactionBody.Builder createUnpause(final UnpauseWrapper unpauseWrapper) {
+        final var builder = TokenUnpauseTransactionBody.newBuilder();
+
+        builder.setToken(unpauseWrapper.token());
+
+        return TransactionBody.newBuilder().setTokenUnpause(builder);
```
Comment (IvanKavaldzhiev) [inline hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/SyntheticTxnFactory.java:440]: You can remove empty lines on 429,431,437,439

File: hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/impl/PausePrecompile.java [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+import java.util.Objects;
+import java.util.function.UnaryOperator;
+import org.apache.tuweni.bytes.Bytes;
+import org.hyperledger.besu.evm.frame.MessageFrame;
+
+public class PausePrecompile extends AbstractWritePrecompile {
+    protected PauseWrapper pauseOp;
+    protected final ContractAliases aliases;
+    protected final EvmSigsVerifier sigsVerifier;
+
```
Comment (IvanKavaldzhiev) [inline hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/impl/PausePrecompile.java:45]: I think `PausePrecompile` won't be inherited, so you can make these fields `private`.

File: hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/impl/UnpausePrecompile.java [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+import com.hederahashgraph.api.proto.java.TransactionBody;
+import java.util.Objects;
+import java.util.function.UnaryOperator;
+import org.apache.tuweni.bytes.Bytes;
+import org.hyperledger.besu.evm.frame.MessageFrame;
+
+public class UnpausePrecompile extends AbstractWritePrecompile {
+    protected UnpauseWrapper unpauseOp;
+    protected final ContractAliases aliases;
+    protected final EvmSigsVerifier sigsVerifier;
```
Comment (IvanKavaldzhiev) [inline hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/impl/UnpausePrecompile.java:44]: I think UnpausePrecompile won't be inherited, so you can make these fields private.

File: test-clients/src/main/java/com/hedera/services/bdd/spec/transactions/TxnVerbs.java [hunk 14]
```diff
@@ -342,6 +342,12 @@ public static HapiContractCall contractCall(
         return new HapiContractCall(abi, contract, params);
     }
 
+    public static HapiContractCall contractCall(
+            String contract, String functionName, String param) {
+        final var abi = getABIFor(FUNCTION, functionName, contract);
+        return new HapiContractCall(abi, contract, param);
+    }
+
```
Comment (IvanKavaldzhiev) [inline test-clients/src/main/java/com/hedera/services/bdd/spec/transactions/TxnVerbs.java:350]: Since `Utils.asHexedAddress` returns `String`, this overloaded method becomes obsolete. It can be removed.

File: test-clients/src/main/java/com/hedera/services/bdd/spec/transactions/TxnVerbs.java [hunk 15]
```diff
@@ -411,6 +417,17 @@ public static HapiContractCreate contractCreate(
...
+    public static HapiContractCreate contractCreate(
+            final String contractName, final String constructorParam) {
+        if (constructorParam != null) {
+            final var constructorABI = getABIFor(CONSTRUCTOR, EMPTY, contractName);
+            return new HapiContractCreate(contractName, constructorABI, constructorParam)
+                    .bytecode(contractName);
+        } else {
+            return new HapiContractCreate(contractName).bytecode(contractName);
+        }
+    }
```
Comment (IvanKavaldzhiev) [inline test-clients/src/main/java/com/hedera/services/bdd/spec/transactions/TxnVerbs.java:429]: Since `Utils.asHexedAddress` returns `String`, this overloaded method becomes obsolete. It can be removed.

File: test-clients/src/main/java/com/hedera/services/bdd/suites/contract/precompile/PauseUnpauseTokenAccountPrecompileSuite.java [hunk 16]
```diff
@@ -0,0 +1,350 @@
...
+        return defaultHapiSpec("PauseFungibleTokenHappyPath")
+                .given(
+                        newKeyNamed(PAUSE_KEY),
+                        cryptoCreate(TOKEN_TREASURY),
+                        cryptoCreate(ACCOUNT).balance(INITIAL_BALANCE),
+                        tokenCreate(VANILLA_TOKEN)
+                                .tokenType(FUNGIBLE_COMMON)
+                                .treasury(TOKEN_TREASURY)
+                                .pauseKey(PAUSE_KEY)
+                                .adminKey(PAUSE_KEY)
```
Comment (IvanKavaldzhiev) [inline test-clients/src/main/java/com/hedera/services/bdd/suites/contract/precompile/PauseUnpauseTokenAccountPrecompileSuite.java:105]: The naming here is a bit confusing. We have `PAUSE_KEY` set to two different types of keys. Probably you can name this key `MULTI_PURPOSE_KEY` or assign `ADMIN_KEY` as admin key.

File: test-clients/src/main/java/com/hedera/services/bdd/suites/contract/precompile/PauseUnpauseTokenAccountPrecompileSuite.java [hunk 17]
```diff
@@ -0,0 +1,350 @@
...
+        return defaultHapiSpec("PauseNonFungibleTokenHappyPath")
+                .given(
+                        newKeyNamed(PAUSE_KEY),
+                        cryptoCreate(TOKEN_TREASURY),
+                        cryptoCreate(ACCOUNT).balance(INITIAL_BALANCE),
+                        tokenCreate(VANILLA_TOKEN)
+                                .tokenType(NON_FUNGIBLE_UNIQUE)
+                                .treasury(TOKEN_TREASURY)
+                                .pauseKey(PAUSE_KEY)
+                                .adminKey(PAUSE_KEY)
```
Comment (IvanKavaldzhiev) [inline test-clients/src/main/java/com/hedera/services/bdd/suites/contract/precompile/PauseUnpauseTokenAccountPrecompileSuite.java:234]: The naming here is a bit confusing. We have `PAUSE_KEY` set to two different types of keys. Probably you can name this key `MULTI_PURPOSE_KEY` or assign `ADMIN_KEY` as admin key.

**Review by IvanKavaldzhiev — APPROVED** [review]
LGTM

File: hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/AbiConstants.java [hunk 18]
```diff
@@ -48,6 +48,10 @@ private AbiConstants() {
     public static final int ABI_ID_DISSOCIATE_TOKENS = 0x78b63918;
     // dissociateToken(address account, address token)
     public static final int ABI_ID_DISSOCIATE_TOKEN = 0x099794e8;
+    // pauseToken(address token)
+    public static final int ABI_PAUSE_TOKEN = 0x7c41ad2c;
+    // unpauseToken(address token)
+    public static final int ABI_UNPAUSE_TOKEN = 0x3b3bff0f;
```
Comment (mustafauzunn) [inline hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/AbiConstants.java:54]: Would you mind changing these to ABI_ID_.... similar to other constants

File: test-clients/src/main/java/com/hedera/services/bdd/suites/contract/precompile/PauseUnpauseTokenAccountPrecompileSuite.java [hunk 19]
```diff
@@ -0,0 +1,349 @@
...
+    }
+
+    @Override
+    protected Logger getResultsLogger() {
+        return log;
+    }
+
+    @Override
+    public List<HapiApiSpec> getSpecsInSuite() {
+        return allOf(
```
Comment (mustafauzunn) [inline test-clients/src/main/java/com/hedera/services/bdd/suites/contract/precompile/PauseUnpauseTokenAccountPrecompileSuite.java:84]: Wouldn't it better to shorten this to:
return List.of(...)

**Review by mustafauzunn — COMMENTED** [review]
LGTM

**Review by Nana-EC — DISMISSED** [review]

**Review by Nana-EC — APPROVED** [review]
LG
Still recommend a shared method in `TxnAwareEvmSigsVerifier` that all the hasActive key methods can use.
